packages/merkle-trees/src/sparse_merkle.nr

```diff
@@ -1,112 +1,208 @@
+        let index_bits: [u1; N] = indexes.to_be_bits();
+
+        // serves as container for hashes and is initialized to be the leaf node
+        let mut node = (self.leaf_hasher)([entry.0, entry.1, T::default()]);
+        // iterates over the list of siblings until the first sibling is found
+        // arbitrarily assigns the sibling to be the left and the node to be the
+        // right element of the hashing pair unless the path indicates the opposite
+        // order in which case the order is changed. The new hash is stored in the container
+        // until the root node is reached and returned.
+        for i in 0..hash_path.len() {
+            let sibling = hash_path[i];
+            if sibling != T::default() {
+                let mut left = sibling;
+                let mut right = node;
+                if index_bits[i] == 0 {
+                    left = node;
+                    right = sibling;
+                }
+                node = (self.hasher)([left, right]);
+            }
+        }
+        node
+    }
+
+    /*
+     * Calculates two roots for a given leaf entry based on the passed array of siblings: one root
+     * for if the leaf entry was included in the tree and one for if the leaf entry was not included
+     * in the tree. This is useful for efficiently proving the membership of leaf entries for a
+     * tree while simultaneously modifying the tree.
+     * @param entry The key and value of an entry [k, v]
+     * @param siblings Contains the siblings from bottom to top
+     * @returns Two root nodes: the first one doesn't include entry, the second does
+     */
+    fn calculate_two_roots<let N: u32>(
+        self,
+        entry: (T, T),
+        indexes: Field,
+        hash_path: [T; N],
+    ) -> (T, T) {
+        let index_bits: [u1; N] = indexes.to_be_bits();
+
+        // root_with_leaf is a container for hashes to derive the root node for the tree that
+        // includes the entry
+        let mut root_with_leaf = (self.leaf_hasher)([entry.0, entry.1, T::default()]);
+        // root_without_leaf is a container for hashes to derive the root node for the tree that
+        // doesn't include the entry
+        let mut root_without_leaf = T::default();
+        // iterate over the levels of the tree from bottom to top
+        for i in 0..hash_path.len() {
+            let sibling = hash_path[i];
+            // After the first sibling is found, the processes are started to calculate the two root nodes.
+            // The calulcation of the root node that includes the entry is comparable to `calculate_root`.
+            // To calc the root node that doesn't include entry, the first sibling is put into the container
+            // and starting from each SUBSEQUENT iteration it is hashed with its sibling and the resulting hash
+            // again stored in the container until the root is reached
+            if sibling != T::default() {
+                if hash_path[i - 1] == T::default() {
```

Comment:
How does this never throw on the first iteration, since `i - 1` would equal `-1` when i equals `0`? Am I missing a certain assumption here?

Reply:
You kind of broke my brain here. But yes, -1 doesn't equal 0 so... It skips this if statement on the first iteration (and correctly so). I would also think it would throw because _if statements are a lie_ in Noir, but the universe works in misterious ways so it doesn't